### ISSUE TITLE
[notabug] initial impressions

### ISSUE THREAD

**Issue by atomGit** [comntr/webext#1]
very interesting and an excellent idea! - i was thinking of something very much like this the last couple of days

1. in order for this to really take off, it has to become very popular and look really good - in that vein, you might want to replace the generic icon for the AMO page with something more elegant

2. you obviously have an interest in respecting privacy - as such, please consider moving the project to GitLab (or some other freedom/privacy respecting repo)

3. in the diagram showing how this ext. networks, it seems to sit between the browser and the destination - i understand you're just grabbing is the comment count apparently, but what happens if comntr.github.io or comntr.live choke? will this slow the loading of the destination site?

4. is it not feasible to add a user name field for commenters? and if that can be added, you might consider changing the 'reply' link to 'reply to [name]' so people don't think they're replying to the whole thread

5. the comments could use some CSS work :) - a dark theme would be great too - i could help with this if desired

6. when one clicks a 'reply' link, the text input should become focused

7. if comments are collapsed, it might be nice to display the number of hidden comments

8. the description for the ext. on AMO is perhaps not as informative as it could be - i can take a stab at editing it if desired

**Comment by comntr** [comment]
I appreciate the attention to my small project :)

1. This is now fixed. Albeit the icon could be more colorful. I'll fix this a bit later.
2. If the extension gains popularity, sure.
3. The extension grabs sha1 of the tab url and asks comntr.live what's the comment count for that sha1.
3.1. Given that comntr.live is a $5/mo VPS, it'll surely choke. In that case the extension will render a "?" over a gray icon. If you try to send a comment, it'll be saved in localStorage/indexedDb and will be sent once comntr.live is back. There is a retry with exp backoff, etc. that keeps sending the unsent comments. User input won't be lost. Now what if comntr.live is taken by someone malicious? The extension has a setting for the "data server url" that can be changed. The grand idea is to have a federated network of such servers, so anyone could start a $5/mo vps, join the network and increase its reliability. But it's too early to invest time there: the public may just appear indifferent to this extension idea. Some time ago I seriously considered ipfs as the backend / distribution mechanism, but had to to give up that idea for numerous reasons.
3.2. If `comntr.github.io` goes down (which is very unlikely), the extension will be able to show the comments count (since it makes requests directly to comntr.live), but will render a blank page if you click on the icon and try to post a comment. For simplicity, the extension's popup simply renders an iframe with comntr.github.io, which introduces a really long delay in slow networks. I'm thinking to have a dead-simple UI in the extension itself, that would render first few comments, have a textbox for adding a comment and a link to the full page. But I thought it's not necessary at this point.
4. That's half way done. The extension uses ed25519 to generate a key pair (the user identity, essentially) and sign every message. The key pair can be wiped if the user chooses so, and the extension will generate a new key pair when the user sends the next comment. Technically, the extension can just attach a user name to every comment and that user name will also be persisted in the local storage.
4.1. The "Reply" is actually to the comment, not to the user. I mean, everybody will see the reply under that comment.
5. If you have time, sure! The main goal is a dead simple UX that will be easy to read (it's not quite there yet) and will be fast to render (10k comments should load momentarily; currently it takes 500 ms). A good example is HN. A bad example is reddit (though it looks nicer).
6. Yep, this needs to be fixed. And the editable comment appearance in general: currently it's hard to figure out that it's something that needs to be sent.
7. Good idea!
8. Please do. I'm about to publish a note about this extension on HN, so I had that audience in mind.

A big missing feature is reply notifications. If you left 50 comments in various places, there won't be an easy way to check if there are any replies. Currently I don't have a good understanding of how to implement this efficiently (keep in mind that it's not Spanner, but a $5/mo vps).

**Comment by atomGit** [comment]
appreciate the detailed answers!

re: point 3 - what happens if comntr.github.io or comntr.live choke? - you served up a great answer, but it wasn't the one i was looking for :) ... let me reword...

let's say we want to visit example.com - if comntr.github.io or comntr.live explode/become overloaded/etc., will the loading time of example.com be affected?

i'll work on the CSS and AMO descrip today very likely

**Comment by atomGit** [comment]
re: CSS  - should i wait until the dust settles to do this? doesn't matter to me at all, i just don't know how solidified stuff is

**Comment by atomGit** [comment]
re: point 4...

> The extension uses ed25519 to generate a key pair (the user identity, essentially) and sign every message. The key pair can be wiped if the user chooses so, and the extension will generate a new key pair when the user sends the next comment. Technically, the extension can just attach a user name to every comment and that user name will also be persisted in the local storage.

not precisely sure how this works or what your intension is, but i would submit that, if this isn't already in the planning stage, every user should be forced to create a permanent and unique user name, whether global or per/domain

i'm thinking here of trolls, people posting personal attacks, a lot of 'F' words, etc. - being assigned a permanent user name might encourage **responsible** free speech - this could also be used in the future to moderate comments, however that is a **very slippery slope** i think and as long as comments can't include anything other than text (i'm thinking of kiddy porn and blatantly illegal shit for example), i wonder if there would ever be a need (legal) to censor anything?

**Comment by comntr** [comment]
The extension shouldn't affect the site in any way. It runs `background.js` that listens to the "tab updated" event, sets a 1 sec timeout and if the tab hasn't changed again during that timeout, it sends a GET to see the number of comments. If the GET fails, the icon simply won't show the number of comments. And comntr.github.io (the UI) isn't involved until the user clicks on the icon.

CSS is sort of stable now. I welcome suggestions how to make the UI more readable. However it's too early to think about a complex UI and I'm not sure if we'll ever need complex UI. I mean, HN works just fine, right? HN even uses a clunky textarea instead of the new contenteditable divs.

*Forcing* users to create user accounts isn't an option, at least for the public version of this idea. I want everyone to be able to make comments no matter what moderators or any sort of authorities have to say.  Otherwise this permanent user account may get blocked over a petty dispute with a moderator. I also want anyone to be able to have as many accounts as they want to. Otherwise a comment left somewhere a year ago may be traced to the actual user identity. Disposable on demand user accounts prevent this kind of tracking. It's not very different from how emails work: we don't have one permanent email address, right?

It's a possibility that there will be a commercial version of this idea. For example, a company may want its employees to comment on something, but it also wants to have control over the discussions. In that case it's fine to have permanent user ids with mods and so on. Or a website admin may want to have comments on his page, but doesn't like the FB solution: he can add a "comntr widget" on the site, configure it to use comntr.foobar.com as the database and force the user identity policy.

As for trolls, I'm generally fine with them. If they want to write something rude, let them do it. If someone feels offended, it's *their* problem, not the trolls'. We obviously can't let the trolls post 10k meaningless comments to every URL, but I believe that the solution is not a permanent user identity, but sophisticated filters on the server (e.g. comntr.live) side. And there will be multiple such servers with different rules. An even better idea is to make creating a valid key pair difficult: we can either require the hash of the public key to have a certain number of leading zeros or ideally to present the user a puzzle that if solved would result in a valid key pair. Now if they start spamming, it'll be easy to remove all their comments and make them produce a new key pair.

There is also content that's strictly prohibited in the US. Since the extension allows only plain text comments under 4 KB, a simple keywords detector + a procedure to address complaints will be enough. But these are just plans and for now I'm going to hide my head in the sand and hope it won't be too bad.

**Comment by remram44** [comment]
Re: 3, you could probably make this more private by using [k-anonymity](https://en.wikipedia.org/wiki/K-anonymity) like [HIBP does](https://blog.cloudflare.com/validating-leaked-passwords-with-k-anonymity/) for their similar use-case. That way, your browser is not sending every single web page you visit in full SHA-1 form, until you actually want to see/post comments.

SHA-1 is not reversible, but that only means that you can't know what page I'm on from the hash. You can totally know if I'm visiting pages from a set you are interested in.

**Comment by comntr** [comment]
I don't quite get the idea. Is it just about sending half of the hash when checking for comments count?

**Comment by remram44** [comment]
Yes pretty much! A low enough number of bits so that it could be any of multiple URLs. The server sends back all the matching hashes with comment counts.

**Comment by comntr** [comment]
I see. I think it's doable - I mean, this shouldn't have a prohibitive perf hit on the server. These hashes have to be organized in a this prefix-tree structure anyway, because otherwise the server will end up with a folder that has 10M subfolders.

**Comment by comntr** [comment]
I've opened #8 to track this.
